keps/sig-node/127-user-namespaces/README.md

```diff
@@ -316,121 +327,108 @@ The picked range will be stored under a file named `userns` in the pod folder
-GetHostIDsForPod() will return nil if the containerUID/containerGID received
-were nil. This is to keep the current functionality untouched, as [fsUser and
-fsGroup can be set to nil][operation-executor] today by the operation executor.
+Let's say we have a pod A that has this mapping for its user namespace
+allocated (same mapping for UIDs than for GIDs):
 
-Secondly, we need to modify these volume types to honor the `fsUser` passed in
-mounterArgs to their SetUpAt() method. As the [AtomicWriter already knows and
-honors the FsUser field][atomic-writer-fsUser] already, just setting this field
-is missing when creating the FileProjection (only done when the feature gate is
-enabled). For example, configmaps will just need to set FsUser alongside Data
-and Mode [here][configmap-fsuser] (and all the other paths in that function that
-create the projection, of course). Other volume types are quite similar too.
+| Host ID  | Container ID   | Length  |
+|----------|----------------|---------|
+| 65536    |      0         |  65536  |
 
-For this we need to modify the MakePayload() or CollectData() functions of each
-supported volume type to handle the fsUser new param, and the corresponding
-volume SetUpAt() function to pass this new param when calling AtomicWriter.
+This means user 0 inside the container is mapped to user 65536 on the host. Let'
+say the pod is running UID 0 inside the container too.
 
-We have written already this code and the per-volume diff is 3 lines.
+##### Example without idmap mounts
 
-For fsGroup() we don't need any changes, these volume types already honors it.
+To understand how idmap mounts work, let's first see an example of what happens
+without idmap mounts.
 
-[volume-mounter-args]: https://github.com/kubernetes/kubernetes/blob/cfd69463deeebfc5ae8a0813d7d2b125033c4aca/pkg/volume/volume.go#L125-L129
-[operation-executor]: https://github.com/kubernetes/kubernetes/blob/cfd69463deeebfc5ae8a0813d7d2b125033c4aca/pkg/volume/util/operationexecutor/operation_generator.go#L674-L675
-[volumeHost interface]: https://github.com/kubernetes/kubernetes/blob/cfd69463deeebfc5ae8a0813d7d2b125033c4aca/pkg/volume/plugins.go#L360-L361
-[atomic-writer-fsUser]: https://github.com/kubernetes/kubernetes/blob/cfd69463deeebfc5ae8a0813d7d2b125033c4aca/pkg/volume/util/atomic_writer.go#L68
-[configmap-fsuser]: https://github.com/kubernetes/kubernetes/blob/cfd69463deeebfc5ae8a0813d7d2b125033c4aca/pkg/volume/configmap/configmap.go#L272-L273
+If such a pod creates a file in any volume, as its mapped to UID 65536 on the
+host, the files it creates will be owned by UID 65536 (i.e. the on disk inode
+will have that value). The pod can read and write to this file just fine. When
+reading the file, the kernel translates the UID 65536 to UID 0 inside the
+container, which is the UID we are running as, and therefore we can read the
+file
 
-#### Phase 2: pods with volumes
+Problems arise if the pod is re-scheduled with another mapping. Let's say it is
```

Comment:
I am not sure how we can hit this issue without volumes. We don't reuse rootfs across pod recreations.

Reply:
Right, brain fart. This example shouldn't be in the stateless pods KEP :D